_posts/2017-12-21-totw-42.md

```diff
@@ -0,0 +1,92 @@
+---
+title: "Tip of the Week #42: Prefer Factory Functions to Initializer Methods"
+layout: tips
+sidenav: side-nav-tips.html
+published: false
+permalink: tips/42
+type: markdown
+order: "042"
+---
+
+Originally posted as totw/42 on 2013-05-10
+
+By Geoffrey Romer [(gromer@google.com)](mailto:gromer@google.com)
+
+Revised 2017-12-21
+
+*"The man who builds a factory builds a temple; the man who works there worships
+there, and to each is due, not scorn and blame, but reverence and praise." --
+Calvin Coolidge*
+
+One of the main difficulties with writing constructors in C++ (at Google) is
+that they effectively must succeed, because they have no way to report failure
+to the caller. (Google's C++ style forbids exceptions.) You can use an
+`abort()`, of course, but doing so crashes the whole program, which is often
+ unacceptable in production code.
+
+If your class’s initialization logic can’t avoid the possibility of failure, one
+common approach is to give the class an initializer method (also called an "init
+method"), which performs any initialization work that might fail, and signals
+that failure via its return value. The assumption is usually that the user will
+call this method immediately after construction, and if it fails the user will
+immediately destroy the object. However, these assumptions are not always
+documented, nor always obeyed. It’s all too easy for users to start calling
+other methods before initialization, or after initialization has failed.
+Sometimes the class actually encourages this behavior, e.g. by providing methods
+to configure the object before initializing it, or to read errors out of it
+after initialization fails.
+
+This design commits you to maintaining a class with at least two distinct
+user-visible states, and often three: initialized, uninitialized, and
+initialization-failed. Making such a design work requires a lot of discipline:
+every method of the class has to specify what states it can be called in, and
+users have to comply with these rules. If this discipline lapses, client
+developers will tend to write whatever code happens to work, regardless of
+what you intended to support. When that starts to happen, maintainability
```

Comment:
hyrumslaw.com ?

Reply:
Done

Reply:
I feel like this should be an honorary Abseil tip ...

Reply:
Yeah.  Blog post maybe.  (That was the original intent, all those years ago.)